Sources/PriorityQueueModule/PriorityQueue.swift

```diff
@@ -225,70 +236,76 @@ public struct PriorityQueue<Element: Comparable> {
       _trickleDownMax(startingAt: index)
     }
   }
-
+ 
+  @inline(__always)
   @inlinable
   internal mutating func _trickleDownMin(startingAt index: Int) {
-    guard let (smallestDescendantIdx, isChild) =
-            _indexOfLowestPriorityChildOrGrandchild(of: index)
-    else {
-      // We have no descendants -- no need to trickle down further
-      return
-    }
+    var index = index
+
+    while true {
+      guard let (smallestDescendantIdx, isChild) =
+              _indexOfLowestPriorityChildOrGrandchild(of: index)
+      else {
+        // We have no descendants -- no need to trickle down further
+        return
+      }
 
-    if isChild {
       if storage[smallestDescendantIdx] < storage[index] {
         _swapAt(smallestDescendantIdx, index)
       }
-    } else {
-      // Smallest is a grandchild
-      if storage[smallestDescendantIdx] < storage[index] {
-        _swapAt(smallestDescendantIdx, index)
-
-        let parentIdx = _parentIndex(of: smallestDescendantIdx)!
-        if storage[smallestDescendantIdx] > storage[parentIdx] {
-          _swapAt(smallestDescendantIdx, parentIdx)
-        }
 
-        _trickleDownMin(startingAt: smallestDescendantIdx)
+      if isChild {
+        return
+      }
+        
+      // Smallest is a grandchild
+      let parentIdx = _parentIndex(of: smallestDescendantIdx)!
+      if storage[smallestDescendantIdx] > storage[parentIdx] {
+        _swapAt(smallestDescendantIdx, parentIdx)
       }
```

Comment:
Doesn't this still need to be wrapped in the `if storage[smallestDescendantIdx] < storage[index]` check? We're potentially doing unnecessary work here, aren't we?

Reply:
Yeah, good catch, mea culpa. I've moved the finishing brace and added a return instead, please have another look now.